java/src/jmri/jmrix/can/CanMessage.java

```diff
@@ -108,125 +127,123 @@ public int hashCode() {
+     * @param a CanMessage or CanReply to test against
      */
     @Override
     public boolean equals(Object a) {
-        if (a == null) {
-            return false;
-        }
-        // check for CanFrame equality, that's sufficient
-        if (a instanceof CanFrame) {
-            CanFrame m = (CanFrame) a;
-            if ((_header != m.getHeader()) || (_isRtr != m.isRtr()) || (_isExtended != m.isExtended())) {
-                return false;
-            }
-            if (_nDataChars != m.getNumDataElements()) {
-                return false;
-            }
-            for (int i = 0; i < _nDataChars; i++) {
-                if (_dataChars[i] != m.getElement(i)) {
-                    return false;
-                }
-            }
-            return true;
-        } else {
-            return false;
-        }
+        return isEqual(this,a);
     }
 
     /**
      * {@inheritDoc}
-     * default toString does not contain the header
-     * this format matches @CanReply
+     * This format matches @CanReply
      */
     @Override
     public String toString() {
-        String s = String.format("[%x] ", _header);
-        for (int i = 0; i < _nDataChars; i++) {
-            if (i != 0) {
-                s += " ";
-            }
-            s = StringUtil.appendTwoHexFromInt(_dataChars[i] & 255, s);
-        }
-        return s;
+        return getToString();
     }
 
+    /**
+     * {@inheritDoc}
+     * This format matches @CanReply
+     */
     @Override
     public String toMonitorString() {
-        StringBuffer buf = new StringBuffer();
-        buf.append("(" + Integer.toHexString(getHeader())
-                + (isExtended() ? " ext)" : ")"));
-        for (int i = 0; i < getNumDataElements(); i++)
-        {
-            buf.append(" " + jmri.util.StringUtil.twoHexFromInt(getElement(i)));
-        }
-	return buf.toString();
+        return monString();
     }
 
+    /**
+     * {@inheritDoc}
+     * Always False
+     */
     @Override
     public boolean replyExpected() {
         return false;
     }
-
-    // accessors to the bulk data
+    
+    /**
+     * {@inheritDoc}
+     */
     @Override
     public int getNumDataElements() {
         return _nDataChars;
     }
 
+    /**
+     * {@inheritDoc}
+     */
     @Override
     public void setNumDataElements(int n) {
         _nDataChars = n;
     }
 
+    /**
+     * {@inheritDoc}
+     */
     @Override
     public int getElement(int n) {
         return _dataChars[n];
     }
 
+    /**
+     * {@inheritDoc}
+     */
     @Override
     public void setElement(int n, int v) {
         _dataChars[n] = v;
     }
 
-    public void setData(int[] d) {
-        int len = (d.length <= 8) ? d.length : 8;
-        for (int i = 0; i < len; i++) {
-            _dataChars[i] = d[i];
-        }
-    }
-
+    /**
+     * Get the data bytes in array form
+     */
     @SuppressFBWarnings(value = "EI_EXPOSE_REP") // OK to expose array, can be directly manipulated
```

Comment:
Since the CanFrame is (it seems from documentation) an Immutable object, while CanMutableFrame is the mutable variant, should we be suppressing an error that manipulation of the internals of the object is possible?

Reply:
I've removed the warning suppression, although I'm still not clear whether direct manipulation is a good or bad thing, there's no CBUS requirement for direct manipulation though there may be other use cases.